### DISCUSSION TITLE
Dev Environment and Debugging

### DISCUSSION THREAD

**Discussion by JrGoodle** [pkgxdev/pkgx#571] [Help]
So I think I've done my due diligence and read most of the wikis, README's, and searched existing discussions; but I haven't found too much info on debugging and dev environment setup.

When I was trying to debug [this issue about @ in package names on the cli](https://github.com/teaxyz/cli/issues/514) I wanted to be able to run a command in VS Code like on the command line, and be able to step through the code. I ended up with [this debug configuration](https://github.com/teaxyz/cli/pull/559), but it doesn't quite work yet as a way to replicate cli commands. I ended up finding a fix mostly through console logs.

While working on packaging up some pantries, I came across a couple that have git submodules. I think in order for those to be built correctly, there's [support needed for git clones instead of downloading tarballs](https://github.com/teaxyz/pantry/issues/762). I started to try to look into what that would take to implement. When I opened brewkit in VS Code, I couldn't get the imports to resolve correctly. The import map is pointing to a github url, but I don't know if there's some trick to getting that to work. I tried adding a modified import map locally, pointing up and outside the brewkit repo to the tea repo. But deno didn't seem to like that. Eventually I realized I could add brewkit inside of the cli repo, and then the imports resolved correctly in VS Code.

I also tried messing around with WebStorm, but I don't think it can read jsonc so I had to change `deno.jsonc` to `deno.json`. And when I moved brewkit inside of cli, WebStorm wouldn't recognize the brewkit directory.

When I moved brewkit inside of cli, I figured I might as well move pantry in there too for convenience. But  once I moved the pantry inside the cli directory, whether I run `pkg build` or `../brewkit/bin/pkg build` I get an error like this:

```
error: Uncaught (in promise) Error: no pantry at path: /Users/jrgoodle/projects/teaxyz/cli, try `tea --sync`
    if (!prefix.exists()) throw new TeaError('not-found: pantry', { path: prefix.parent() })
                                ^
    at entry (https://raw.githubusercontent.com/teaxyz/cli/v0.30/src/hooks/usePantry.ts:149:33)
    at Object.available (https://raw.githubusercontent.com/teaxyz/cli/v0.30/src/hooks/usePantry.ts:140:29)
    at file:///Users/jrgoodle/projects/teaxyz/cli/brewkit/libexec/available.ts:15:21
    at eventLoopTick (ext:core/01_core.js:181:11)
warning: not supported on current platform
```

So I moved the pantry back out and then the `pkg` commands worked with brewkit inside of the cli repo. But at this point I realized there's no obvious way I could find to have the brewkit repo call into the code in the cli repo, it always goes out to the tea installation. This seems like it would be pretty convenient to be able to seamlessly step between brewkit code and cli code. I know I can install the current version of the cli code to my tea installation, but that's more overhead and way less interesting than being able to step through the code in a debugger.

I started trying to mess around with adding VS Code launch configurations for running `pkg` commands, but they start as bash scripts so there's probably some environment setup required to be able to call directly into the .ts files correctly.

Sorry if that's mostly rambling. I just wanted to describe most of what I've tried so far. And I guess my question is, what do the tea team members' workflows look like when hacking on brewkit and cli? What's the state of the art for debugging currently?

**Comment by mxcl** [comment]
We added git clones recently, via git+https urls, see pantry for deets.

Dunno what webstorm is.
